tasks/init.rb

```diff
@@ -6,10 +6,10 @@
 
 def get(fact)
   if ENV.key? 'Path'
-    facter = '"C:\Program Files\Puppet Labs\Puppet\bin\facter"'
+    facter = 'C:\Program Files\Puppet Labs\Puppet\bin\facter'
```

Comment:
Never assume root drive is C. Don't assume InstallPath is 'Program Files'. This is brittle code that will fail

Reply:
Yes. I'd like to see it improved. However it does walk PATH. That was what someone came up with last time I pointed this out and said we should query the registry.

Reply:
Yeah, `PATH` walking should generally find what you're looking for, so this fallback hopefully never gets hit anyhow... but you could at least be a little more resilient against a relocated Program Files directory with:

```ruby
if File.exists?("#{ENV['ProgramFiles']}\\Puppet Labs\\Puppet\\bin\\facter")
  facter = "#{ENV['ProgramFiles']}\\Puppet Labs\\Puppet\\bin\\facter"
end
```

If you'd rather just use the registry to find the installation location rather than assuming `Program Files`, this will look for 64-bit key and fallback to 32-bit, crashing and burning if the keys don't exit:

```ruby
require 'win32/registry'

Win32::Registry::HKEY_LOCAL_MACHINE.open('SOFTWARE\Puppet Labs\Puppet') do |reg| 
  installed_dir = reg['RememberedInstallDir64'] rescue nil
  installed_dir = reg['RememberedInstallDir'] if installed_dir.nil?
end
```

Reply:
Is that a request to change it? That means the gem will depend on `win32/registry` unless I rewrite the registry support myself. How bad would shelling out to powershell be instead?

Eventually I see this being subsumed by https://github.com/puppetlabs/bolt/blob/master/modules/facts/tasks/powershell.ps1 which will need its own updates, but I think it's a ways off before we deprecate facter_task in favor of a new facts module.

Reply:
Fortunately `win32/registry` is part of the core Ruby stdlib (at least as far back as [1.8.7](https://ruby-doc.org/stdlib-1.8.7/libdoc/Win32API/rdoc/Win32/Registry.html), so you don't have any additional dependencies to pull in to use it.

What I gave you was just a quick example, but should mostly work as-is. Might want to add an additional check like this:

```ruby
Win32::Registry::HKEY_LOCAL_MACHINE.open('SOFTWARE\Puppet Labs\Puppet') do |reg| 
  installed_dir = reg['RememberedInstallDir64'] rescue nil
  break if FileUtils.exist?(installed_dir)
  installed_dir = reg['RememberedInstallDir']
end
```

I think we only care about 64-bit on 64-bit OS and 32-bit on 32-bit OS, and I believe this should work... but might want to manually vet that to be absolutely certain.